[숫자 야구 게임] 복수한다!!! 미션 제출합니다.

snowman95 opened the pull request:
안녕하세요. 
우테코 복수방 복수한다!!! 입니다.

snowman95 committed:
Create README.md

docs(README.md): 프로젝트 설명 및 요구 사항 작성

프로젝트 설명 및 요구 사항 전체 내용 작성
README.md 파일에 명시된 "기능 요구 사항" 단위로 Commit 예정

snowman95 committed:
feat(gameStart, play): 게임 시작 및 초기화, 입력 비교 함수 생성

gameStart 의 신규 기능:
- 게임 초기화 (상대방(컴퓨터)가 1~9 중 서로 다른 임의의 수 3개 선택)

play의 신규 기능:
- 사용자가 입력창에 입력한 3자리 수를 받아서 상대방의 숫자와 비교

게임과 직접 관련된 함수만 index.js에 작성
Util 함수는 tools.js 파일에 따로 작성하여 import 후 사용

snowman95 committed:
feat(submitButton): 확인 버튼 클릭 시 결과 출력

index.js 의 신규 기능:
- 확인 버튼 클릭 : 사용자가 입력한 숫자와 상대방의 숫자와 비교하여 힌트 출력

snowman95 committed:
feat(start, clear): 게임 초기화 로직 추가, 게임 클리어 추가

start의 신규 기능:
- 사용자 입력 초기화
- 게임 결과 텍스트 초기화

clear의 신규 기능:
- 종료 조건 (3스트라이크) 충족 시 게임 종료 문구 및 재시작 버튼 출력
- 재시작 버튼 클릭 시 게임 재시작

함수명 변경 : gamePlay → play
버튼 이벤트 등록 함수를 BaseballGame 내부로 이동

snowman95 committed:
refector(match): 볼/스트라이크/낫싱 문자열을 바로 반환하도록 변경

match의 변경점:
- matchBalls로 이름 변경
- ball, strike 의 개수를 반환하지 않고 결과 문자열로 반환하도록 변경

 Before:

    return: {
      matchInclude, matchExact
    }

 After

    return: "n볼n스트라이크" or "낫싱"

snowman95 committed:
feat(checkValid): 사용자 입력 유효성 검사

checkValid의 신규 기능:
- 사용자 입력 값의 유효성을 검사하여 true/false 반환
- 사용자가 잘못된 값 입력 시 에러 메시지 alert

snowman95 committed:
fix(createRandNum,matchBalls): 랜덤함수 변경 및 출력 문구 수정

createRandNum의 fix 내용 :
- 기능 요구 사항과 알맞지 않은 pickUniqueNumbersInRange 함수를 제거하고 pickNumberInRange를 사용하도록 수정함.

matchBalls의 fix 내용 :
- 출력 문구가 x볼x스트라이크의 경우 한칸 띄워서 x볼 x스트라이크로 출력 되도록 변경

Tanney-102 commented on src/tools.js:
```diff
@@ -0,0 +1,91 @@
+    newRandomNum.add(MissionUtils.Random.pickNumberInRange(1, 10));
+  }
+  return [...newRandomNum];
+};
```
함수의 역할을 설명하기 위해 주석을 적어주는 것도 좋습니다. 그러나 주석을 달기 전에는 항상 꼭 필요한 주석인지 고민할 필요가 있습니다. 위 주석에서 첫번째 줄(1~9중 서로 다른...)은 함수나 인자의 이름으로 드러내야할 부분이지 않을까요?
예를들면 
```js
const getDistinctRandomNumbers(min, max, length) {}
```
이런식로요!

snowman95 replied:
네 말씀해주신 내용 충분히 이해했습니다.
이게 과제라고 생각하다 보니 사실 불필요할 정도로 많은 주석이 추가된 것 같습니다.
하나의 개인 프로젝트로 받아들이고 접근해보겠습니다.

(여담)
제가 원래는 함수/변수명을 너무 길지 않은 선에서 읽으면 기능을 이해할 수 있도록 충분히 길게 명칭을 짓는 편인데
프로그래밍 요구사항에 있는 함수 길이 15줄에 너무 신경을 쓰다보니 도중에 기억이 왜곡되어버려서 함수명도 15자 이내로 짓고 있더군요. 
제한 사항을 지켜야 되는 것은 정말 힘든 것 같습니다...

Tanney-102 commented on src/tools.js:
```diff
@@ -0,0 +1,91 @@
+    newRandomNum.add(MissionUtils.Random.pickNumberInRange(1, 10));
+  }
+  return [...newRandomNum];
+};
```
위 코드에서의 3, 1, 10과 같은 숫자들은 가능한 의미있는 식별자에 할당해서 사용하는 것이 좋습니다.
예)
```js
// src/constants/baseball.js
const BASEBALL_GAME_INPUT = {
    LENGTH: 3,
    MIN: 1,
    MAX: 10,
};
```

Tanney-102 replied:
중복을 제거하는 방법으로 Set을 사용하고 계시네요!
그런데 위와 같은 방법으로는 연산량을 보장받을 수 없지 않을까요?
물론 숫자 3개정도 뽑는데 있어서 크게 이슈는 없겠지만요 ㅎㅎ
(이부분은 그냥 개인적인 생각입니다.)

snowman95 replied:
(프로그래밍 요구사항 - MissionUtils 라이브러리의 Random.pickNumberInRange를 사용하여 랜덤값 가져온다.)
위 프로그래밍 요구사항을 준수하려다 보니 더 효율적으로 짜기가 쉽지 않았네요 ㅠㅠ

요구사항 제한이 없었다면 해당 라이브러리에 pickUniqueNumbersInRange 라는 함수를 쓰거나 직접 구현했을 듯 합니다. 😅😅

Tanney-102 commented on src/tools.js:
```diff
@@ -0,0 +1,91 @@
+    valid = true;
+  }
+
+  if (valid === false) alert(message);
```
함수의 이름은 checkValid인데 유효성 검사 이후 그 결과를 '처리'하는 것까지 맡아서 하고 있네요!
유효하지 않은 값에 대해 alert으로 처리할지 snackbar로 처리할지 등... 함수를 사용하는 측에 맡겨야하지 않을까요?

snowman95 replied:
네 맞습니다. true/false 만 반환하는 함수였으나... 여러 선택과 고민 사이에서
결국 checkValid 함수에 alert를 넣는 선택을 하게  되었습니다.

[머릿속 선택과 고민]
Q. play 함수 내부에서 if문을 쓸 수 있는가?          → 아니요. depth 요구조건을 어기게 됨
Q. if문을 쓸 수 없으니 삼항 연산자를 쓸 것인가? → 아니요. 코드가 너무 지저분해짐


피드백 이후에 삼항 연산자를 쓰는 방법으로 변경했는데 
생각보다 그렇게 지저분하지는 않네요.
```
const { valid, message } = checkValid(userInput);
    let result = "";
    valid ? (result = matchBalls(computerRandNum, userInput)) : alert(message);
    return result;
```

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,59 @@
+    inputText.value = "";
+    resultDiv.innerHTML = "";
+    randomNum = createRandNum();
+    console.log(randomNum);
```
console.log는 지워도 될 것 같습니다.

snowman95 replied:
동의합니다.

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,59 @@
+  };
+
+  // 게임 시작
+  this.start();
```
constructor 안에서 해야할 일일까요?
```js
const baseballGame = new BaseballGame();

baseballGame.start();
```
이런식의 사용이 더 객체답지 않을까요?

snowman95 replied:
동의 합니다.

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,59 @@
+    this.start();
+  });
+}
+
```
이벤트 바인딩도 constructor에서 할 일은 아닌 것 같습니다.

snowman95 replied:
동의합니다.

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,59 @@
+restartButton.id = "game-restart-button";
+
+export default function BaseballGame() {
+  let randomNum = [];
```
constructor안에서 정의하는 '상태'라면 this.randomNum = []이 더 적절하지 않을까요?

snowman95 replied:
자바스크립트 공부를 소홀히 했던 게 딱 걸린 것 같네요. 😂
class 랑 생성자 함수 부분을 더 공부를 더 해야겠습니다...

피드백 주신 대로 function 대신 class로 구현하였고
class의 constructor 내부에서 초기화 이루어지도록 변경했습니다.

Tanney-102 commented on src/tools.js:
```diff
@@ -0,0 +1,91 @@
+    : "낫싱";
+};
+
+export function showGameResult(hint, resultDiv, restartBtn) {
```
화살표함수와 function키워드가 혼용되고 있네요.
한가지로 통일하면 좋겠습니다.

snowman95 replied:
화살표 함수 vs function 키워드
어느 쪽을 사용할 것인지 보통 팀 내에서 규칙을 정해 놓고 따르는 방식으로 작업 하시나요?
이 부분이 항상 의문이긴 했습니다.

Tanney-102 commented on src/index.js:
```diff
@@ -0,0 +1,59 @@
+const restartButton = document.createElement("button");
+
+restartButton.innerHTML = "게임 재시작";
+restartButton.id = "game-restart-button";
```
createElement는 사용하는 함수 안에서(showGameResult) 하는게 좋겠습니다.

snowman95 replied:
동의합니다.

Tanney-102 reviewed (changes requested):
안녕하세요 snowman님! 우아한테크코스 3기 브콜입니다.
빠르게 구현해주셨네요~ㅎㅎ 
몇가지 피드백 남깁니다!

1. 객체 지향 프로그래밍과 constructor의  역할에 대해 공부해보시면 좋을 것 같아요. 지금으로서는 BaseballGame이라는 함수가 일반함수인지 생성자인지 역할이 모호한 것 같습니다.
2. function보다는 class 문법을 고려해보시면 좋을 것 같아요. 고려하시는 김에 과거 function 키워드로 함수를 선언하던 때와 ES6 이후 추가된 함수 사양들 사이에 어떤 차이점이 있는지도 공부해보시면 좋을 것 같습니다.

더 자세한 내용은 코멘트로 남겼습니다!

snowman95 committed:
docs(README.md): PR#1 피드백 사항 작성

Pull Request 후 받은 PR#1 피드백 내용을 README.md에 정리

snowman95 committed:
refactor: PR#1 피드백 사항 반영

PR#1 피드백 사항 반영 :
- function 키워드와 화살표 함수 중 하나로 통일
- 주석보다 함수명과 인자를 통해 역할을 드러내기
- 상수를 의미있는 식별자로 표현하기 (LENGTH, MIN, MAX)
- checkValid 함수가 한 가지 역할만 하도록 변경 → 결과 처리 부분은 따로 하도록
console.log() 제거
- class constructor안에서 이벤트 바인딩, Element 생성하도록 변경
- class 객체 생성 후 시작함수 호출

snowman95 committed:
style(tools): 화살표 함수를 function 키워드로 통일

tools 파일 내 화살표 함수를 function 키워드로 통일

snowman95 committed:
refactor(play, checkValid): 유효성 검사 함수 반환 값 수정

checkValid 의 수정 사항:
- 기존 valid(bool), message(text) 로 반환하던 것을 message(text) 하나만 반환
- 반환 값이 존재하면 프로그램 이어서 실행되도록 수정

snowman95 commented:
git page : https://snowman95.github.io/javascript-baseball-precourse-1/

PR#1 피드백 반영 완료했습니다. 정말로 감사드립니다. 😊
